### ISSUE TITLE
Support drag and drop with nested lists

### ISSUE THREAD

**Issue by psycry** [hanshengchiu/reorderables#2]
There may be a way to do this I'm overlooking, but whenever I nest wraps inside warps (or lists) and attempt to drag and drop the copying function seems to throw duplicate global key errors when I pick up the object.  The alternative is a sub-list that doesn't save state which isn't ideal.  I've tried some ways around this but it looks like the uuid library I'm using is possibly duplicating values or there's something in the way the copy function works that the tree has duplicate keys when moving the items.

So in other words I have a list of tiles with sub-lists, whenever I try to move them it throws an exception.

**Comment by hanshengchiu** [comment]
Can you show some code of how this duplicate global key error happens?

**Comment by psycry** [comment]
Sure, I just kind of made this quickly but say you have 2 widgets with reorderableWraps, if you put the second inside the first and try to drag it doesn't handle the dragging and throws a key error.  I'm using these so that you can grow the list, it works fine on the top level but once you add the second it starts having issues with the conflicting keys, I'm assuming because of the way it's switching them out.  I may just have a bug in my code but I've tried this a few ways and it's only on the second nesting it starts having issues.  It's either do it like this, or the state of the sub-list gets lost after a drag.  This should reproduce it:

`
class WrapExample extends StatefulWidget {
  WrapExample() : super(key: ValueKey(Uuid().v1().toString()));
  final List<Widget> _exampleList = List();
  @override
  State<StatefulWidget> createState() {
    return _WrapExampleState();
  }
}

class _WrapExampleState extends State<WrapExample> {
//  final double _iconSize = 90;

  @override
  Widget build(BuildContext context) {
    return ListView(shrinkWrap: true, primary: false, children: [
      ReorderableWrap(
          key: ValueKey(Uuid().v1().toString()),
          spacing: 8.0,
          runSpacing: 4.0,
          padding: const EdgeInsets.all(8),
          children: widget._exampleList,
          onReorder: _onReorder),
      Row(mainAxisAlignment: MainAxisAlignment.end, children: [
        FloatingActionButton(
          key: ValueKey(Uuid().v1().toString()),
          onPressed: () {
            setState(() {
              widget._exampleList.add(Wrap2Example());

//              currentPageIndex = PageIndex.ITINERARY.index;
            });
          },
          tooltip: 'Add item',
          child: Icon(Icons.add),
        ),
      ]),
    ]);
  }

  void _onReorder(int oldIndex, int newIndex) {
    widget._exampleList
        .insert(newIndex, widget._exampleList.removeAt(oldIndex));
//    });
  }
}
`

`
class Wrap2Example extends StatefulWidget {
  Wrap2Example() : super(key: ValueKey(Uuid().v1().toString()));
  final List<Widget> _exampleList = List();
  @override
  State<StatefulWidget> createState() {
    return _Wrap2ExampleState();
  }
}

class _Wrap2ExampleState extends State<Wrap2Example> {
//  final double _iconSize = 90;

  @override
  Widget build(BuildContext context) {
    return ListView(shrinkWrap: true, primary: false, children: [
      ReorderableWrap(
          key: ValueKey(Uuid().v1().toString()),
          spacing: 8.0,
          runSpacing: 4.0,
          padding: const EdgeInsets.all(8),
          children: widget._exampleList,
          onReorder: _onReorder),
      Row(mainAxisAlignment: MainAxisAlignment.end, children: [
        FloatingActionButton(
          key: ValueKey(Uuid().v1().toString()),
          onPressed: () {
            setState(() {
              widget._exampleList.add(Wrap2Example());

//              currentPageIndex = PageIndex.ITINERARY.index;
            });
          },
          tooltip: 'Add item',
          child: Icon(Icons.add),
        ),
      ]),
    ]);
  }

  void _onReorder(int oldIndex, int newIndex) {
    widget._exampleList
        .insert(newIndex, widget._exampleList.removeAt(oldIndex));
//    });
  }
}
`

**Comment by psycry** [comment]
Sorry about the formatting it's not working correctly for some reason when I try to add it in a code block.

**Comment by app-o-matix** [comment]
@psycry 
You need triple backticks for blocks of code. Single back ticks are for a single line or less. The code at the end that is formatting is doing so because indenting 4 spaces also works.

```dart
class WrapExample extends StatefulWidget {
  WrapExample() : super(key: ValueKey(Uuid().v1().toString()));
  final List _exampleList = List();
  @override
  State createState() {
    return _WrapExampleState();
  }
}

class _WrapExampleState extends State {

  @override
  Widget build(BuildContext context) {
    return ListView(shrinkWrap: true, primary: false, children: [
      ReorderableWrap(
          key: ValueKey(Uuid().v1().toString()),
          spacing: 8.0,
          runSpacing: 4.0,
          padding: const EdgeInsets.all(8),
          children: widget._exampleList,
          onReorder: _onReorder),
      Row(mainAxisAlignment: MainAxisAlignment.end, children: [
        FloatingActionButton(
          key: ValueKey(Uuid().v1().toString()),
          onPressed: () {
            setState(() {
              widget._exampleList.add(Wrap2Example());
            });
          },
          tooltip: 'Add item',
          child: Icon(Icons.add),
        ),
      ]),
    ]);
  }

  void _onReorder(int oldIndex, int newIndex) {
    widget._exampleList
        .insert(newIndex, widget._exampleList.removeAt(oldIndex));
  }
}

class Wrap2Example extends StatefulWidget {
  Wrap2Example() : super(key: ValueKey(Uuid().v1().toString()));
  final List _exampleList = List();
  @override
  State createState() {
    return _Wrap2ExampleState();
  }
}

class _Wrap2ExampleState extends State {

  @override
  Widget build(BuildContext context) {
    return ListView(shrinkWrap: true, primary: false, children: [
      ReorderableWrap(
          key: ValueKey(Uuid().v1().toString()),
          spacing: 8.0,
          runSpacing: 4.0,
          padding: const EdgeInsets.all(8),
          children: widget._exampleList,
          onReorder: _onReorder),
      Row(mainAxisAlignment: MainAxisAlignment.end, children: [
        FloatingActionButton(
          key: ValueKey(Uuid().v1().toString()),
          onPressed: () {
            setState(() {
              widget._exampleList.add(Wrap2Example());

            });
          },
          tooltip: 'Add item',
          child: Icon(Icons.add),
        ),
      ]),
    ]);

  }

  void _onReorder(int oldIndex, int newIndex) {
    widget._exampleList
        .insert(newIndex, widget._exampleList.removeAt(oldIndex));
  }
} 
```

**Comment by hanshengchiu** [comment]
this issue has just been fixed in v0.2.5 [pub](https://pub.dartlang.org/packages/reorderables/versions/0.2.5) An example of nesting wraps was also added.

[screenshot](https://github.com/hanshengchiu/reorderables/blob/master/example/gifs/nested_reorderable_wrap_small.gif?raw=true)

**Comment by psycry** [comment]
Fixed code formatting (er... sort of, thanks for the tripple backtick information, the indent is probably still conflicting but since you copied it no point in updating any further).

Tested and can confirm this is working now at least at the nesting level I found it, I'm having some state issues but I probably need to change the code I had, was out of the country so didn't see my emails when you fixed this so sorry for the delayed response, thanks for the work put into this on the fix.
